### PR TITLE
[autoscaler] Add support for multiple cluster per Azure resource group

### LINKED ISSUE #22996: [Bug] [Autoscaler] [Azure] Can't create multiple clusters on Azure
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Clusters

### Issue Severity

Medium: It contributes to significant difficulty to complete my task but I work arounds and get it resolved.

### What happened + What you expected to happen

Ray autoscaler on Azure couldn't create multiple clusters in the same resource group.

After successfully create a cluster on Azure and keep it running,  then edit the `cluster_name` in the yaml and run the same command to create a new cluster in the same resource group. It will update the existing cluster, instead of create the new cluster.
```
ray up ray/python/ray/autoscaler/azure/example-full.yaml
```

### Versions / Dependencies

Ray-1.11.0

### Reproduction script

```
ray up ray/python/ray/autoscaler/azure/example-full.yaml
# after the cluster is successfully created and running
# then edit the cluster_name of yaml to create a new cluster
vim ray/python/ray/autoscaler/azure/example-full.yaml
ray up ray/python/ray/autoscaler/azure/example-full.yaml
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by gramhagen**
## Why are these changes needed?

Adding support for deploying multiple clusters into the same azure resource group

Changes:
- Adding unique_id field to provider section of yaml, if not provided one will be created based on hashing the resource group and cluster name. This will be appended to the name of all resources deployed to azure so they can co-exist in the same resource group (provided the cluster name is changed)
- Pulled in changes from #22997 to use cluster name when filtering vms to retrieve nodes only in the current cluster
- Added option to explicitly specify the subnet mask, otherwise use the resource group and cluster name as a seed and randomly choose a subnet to use for the vnet (to avoid collisions with other vnets)
- Updated yaml example files with new provider values and explanations
- Pulling resource_ids from initial azure-config-template deployment to pass into vm deployment to avoid matching hard-coded resource names across templates

## Related issue number

Closes #22996 
Supersedes #22997 

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(
   - [X] Manual tests

### CODE REVIEW

File: python/ray/autoscaler/_private/_azure/config.py
Comment (wuisawesome) [inline python/ray/autoscaler/_private/_azure/config.py:93]: should we seed this with `unique_id`? I imagine we want to keep the same subnets if the unique_id is the same?

File: python/ray/autoscaler/_private/_azure/config.py
Comment (wuisawesome) [inline python/ray/autoscaler/_private/_azure/config.py:85]: What's the use case for `unique_id`? I imagine it's primarily if you want to change the cluster name without losing its identity? In that case, should we log the id?

File: python/ray/autoscaler/_private/_azure/azure-config-template.json
Comment (wuisawesome) [inline python/ray/autoscaler/_private/_azure/azure-config-template.json:8]: Can you mention why we'd want to set this, and what it defaults to?

File: python/ray/autoscaler/azure/example-full.yaml
Comment (wuisawesome) [inline python/ray/autoscaler/azure/example-full.yaml:47]: Same as in the schema/template above, can you provide a little more detail about how we derive it? e.g. `If not set, the default unique id will be derived from the cluster name and resource group.`

**Review by wuisawesome — CHANGES_REQUESTED** [review]
The high level approach makes sense to me (though I don't know a whole lot about azure). Comments/questions are mostly about polish. 

Thanks for cleaning this up!

File: python/ray/autoscaler/_private/_azure/azure-config-template.json
Comment (gramhagen) [reply python/ray/autoscaler/_private/_azure/azure-config-template.json:8]: will do, though there is no default here, that is set prior to exercising this template.

File: python/ray/autoscaler/_private/_azure/config.py
Comment (gramhagen) [reply python/ray/autoscaler/_private/_azure/config.py:85]: this is just to be able to isolate resources in azure based on the cluster, so they can live in the same resource group and not conflict. i did not want to rely on the cluster name itself because there are restrictions on characters and size of resource names, so i just hash it to create something mostly unique.

File: python/ray/autoscaler/_private/_azure/config.py
Comment (gramhagen) [reply python/ray/autoscaler/_private/_azure/config.py:93]: yeah, that would be easier, will change

**Comment by gramhagen** [comment]
> The high level approach makes sense to me (though I don't know a whole lot about azure). Comments/questions are mostly about polish.
> 
> Thanks for cleaning this up!

thanks for the review! let me know if there are any additional changes that can improve this.

File: python/ray/autoscaler/_private/_azure/config.py
Comment (wuisawesome) [reply python/ray/autoscaler/_private/_azure/config.py:85]: Hmmm I see. If we want to just use it as an implementation detail, should we remove the cluster yaml entirely? 

If we want to treat it as an implementation detail, we probably shouldn't include it in the cluster yaml API at all. If we do want to expose it, I think we should probably INFO log as a matter of principle (logging the default values we fill in for explicitely exposed fields in general).

I defer to you on whether unique_id should be an implementation detail or a documented part of the node provider.

Comment (gramhagen) [reply python/ray/autoscaler/_private/_azure/config.py:85]: i think it makes sense to log it, let me know if the latest update is what you had in mind

**Review by wuisawesome — APPROVED** [review]

**Comment by wuisawesome** [comment]
Thanks for all the polish fixes here. I appreciate it! Will merge once tests pass.

**Comment by wuisawesome** [comment]
Hey @gramhagen do you mind rebasing this? hopefully CI should be less flakey now.

**Comment by gramhagen** [comment]
> Contributor

done

**Comment by wuisawesome** [comment]
Test failures (gcs ha, rllib, serve) look unrelated. Linkcheck was broken in master, merging.

**Comment by wuisawesome** [comment]
@gramhagen sorry one final thing, can you do the DCO sign off? (you can just use the DCO toggle in the UI if you want)

**Comment by gramhagen** [comment]
ugh...

**Comment by gramhagen** [comment]
@wuisawesome I don't use a UI for git and when following the instructions in the DCO check it overwrote every commit from when I previously rebased master. Is there an easier way you would suggest signing just my commits from the commandline?

**Comment by cadedaniel** [comment]
> @wuisawesome I don't use a UI for git and when following the instructions in the DCO check it overwrote every commit from when I previously rebased master. Is there an easier way you would suggest signing just my commits from the commandline?

This is still signing commits from the command line, but with some automation:
```bash
git rebase -i -x "git commit --amend --no-edit --signoff" HEAD~3
```

This will open your editor with the `git-rebase-todo` file; remove the `exec ` line for all commits from the merge commit(s).

**Comment by gramhagen** [comment]
this is not working too well... if it's a hard requirement it will be easier to just create a new branch with a single signed commit

**Comment by wuisawesome** [comment]
@gramhagen i mean you should be able to toggle it from this link https://github.com/ray-project/ray/pull/26392/checks?check_run_id=7845456608

**Comment by gramhagen** [comment]
i don't see any kind of toggle?
image

**Comment by wuisawesome** [comment]
@gramhagen it seems you there's no button because you aren't a member of the rayproject github org, you were just sent an invite, can you try now?

**Comment by gramhagen** [comment]
i gave up and cherry-picked everything into a new branch, squashed it, signed it, and force pushed it. i love git, until i don't...

**Pull request merged**
